### PR TITLE
OS Login and Linux SSH Key managers

### PR DESCRIPTION
**Pull request by hopkiw**
This is a translation of the existing `google_accounts_daemon` python script. It moves the existing `accountsMgr` to `winAccountsMgr` and introduces the `linuxAccountsMgr` and `osloginMgr` managers.

### CODE REVIEW

File: GCEWindowsAgent/diagnostics.go [hunk 1]
```diff
@@ -27,8 +26,7 @@ import (
 const diagnosticsCmd = `C:\Program Files\Google\Compute Engine\diagnostics\diagnostics.exe`
 
 var (
-	diagnosticsRegKey   = "Diagnostics"
```
Comment (adjackura) [inline GCEWindowsAgent/diagnostics.go]: Since your here anyway can you just drop this manager completly, if you want to move this to another PR thats OK to

File: GCEWindowsAgent/main.go [hunk 2]
```diff
@@ -37,6 +37,7 @@ var (
 	ticker                   = time.Tick(70 * time.Second)
 	oldMetadata, newMetadata *metadataJSON
 	config                   *ini.File
+	osrelease                release
```
Comment (adjackura) [inline GCEWindowsAgent/main.go]: osRelease

File: GCEWindowsAgent/oslogin.go [hunk 3]
```diff
@@ -0,0 +1,339 @@
...
+func (a *osloginMgr) timeout() bool {
+	return false
+}
+
+func (a *osloginMgr) disabled(os string) bool {
+	return os == "windows"
+}
+
+func (a *osloginMgr) set() error {
+	oldenable := oldMetadata.Instance.Attributes.EnableOSLogin || oldMetadata.Project.Attributes.EnableOSLogin
```
Comment (adjackura) [inline GCEWindowsAgent/oslogin.go]: Use camelCase when you have two distinct words.

https://golang.org/doc/effective_go.html#mixed-caps

File: GCEWindowsAgent/oslogin.go [hunk 4]
```diff
@@ -0,0 +1,339 @@
...
+
+	if enable && !oldenable {
+		logger.Infof("Enabling OS Login")
+		newMetadata.Instance.Attributes.SSHKeys = nil
+		newMetadata.Project.Attributes.SSHKeys = nil
+		(&linuxAccountsMgr{}).set()
+	}
+
+	if err := updateSSHConfig(enable, twofactor); err != nil {
+		logger.Errorf("error updating SSH config: %v\n", err)
```
Comment (adjackura) [inline GCEWindowsAgent/oslogin.go]: Log lines should be capitalized and end in punctuation.

File: GCEWindowsAgent/oslogin.go [hunk 5]
```diff
@@ -0,0 +1,339 @@
...
+			logger.Errorf("error restarting service: %v\n", err)
+		}
+	}
+
+	if enable {
+		if err := exec.Command("google_oslogin_nss_cache").Run(); err != nil {
+			logger.Errorf("error updating NSS cache: %v\n", err)
+		}
+	}
+	os.Exit(1)
```
Comment (adjackura) [inline GCEWindowsAgent/oslogin.go]: I assume this was a debug line?

File: GCEWindowsAgent/oslogin.go [hunk 6]
```diff
@@ -0,0 +1,339 @@
...
+
+	return nil
+}
+
+func filterGoogleLines(contents string) []string {
+	var isgoogle, isgoogleblock bool
+	var filtered []string
+	for _, line := range strings.Split(contents, "\n") {
+		if strings.Contains(line, googleComment) {
```
Comment (adjackura) [inline GCEWindowsAgent/oslogin.go]: A switch statement with a default of filtered = append(filtered, line) would be easier to read and drop the need for all the continues

File: GCEWindowsAgent/oslogin.go [hunk 7]
```diff
@@ -0,0 +1,339 @@
...
+		osloginSudoers = "/usr/local/etc/sudoers.d/google-oslogin"
+	}
+	sudoFile, err := os.OpenFile(osloginSudoers, os.O_WRONLY|os.O_CREATE|os.O_EXCL, 0440)
+	if err != nil {
+		if os.IsExist(err) {
+			return nil
+		}
+		return err
+	}
+	defer sudoFile.Close()
```
Comment (adjackura) [inline GCEWindowsAgent/oslogin.go]: check the error, for this its simple to just return sudoFile.Close() as you have no other returns

File: GCEWindowsAgent/windowsaccounts.go [hunk 8]
```diff
@@ -0,0 +1,288 @@
+//  Copyright 2017 Google Inc. All Rights Reserved.
```
Comment (adjackura) [inline GCEWindowsAgent/windowsaccounts.go]: Rename to windows_accounts.go

File: GCEWindowsAgent/oslogin.go [hunk 9]
```diff
@@ -0,0 +1,339 @@
+//  Copyright 2019 Google Inc. All Rights Reserved.
```
Comment (adjackura) [inline GCEWindowsAgent/oslogin.go]: os_login.go may make more sense, but I don't mind of you keep this as oslogin.go

File: GCEWindowsAgent/windowsaccounts.go (hunk 8 above)
Comment (hopkiw) [reply GCEWindowsAgent/windowsaccounts.go]: and later, windows_accounts_windows.go :)

File: GCEWindowsAgent/oslogin.go [hunk 10]
```diff
@@ -0,0 +1,329 @@
...
+
+	if enable && !oldEnable {
+		logger.Infof("Enabling OS Login")
+		newMetadata.Instance.Attributes.SSHKeys = nil
+		newMetadata.Project.Attributes.SSHKeys = nil
+		(&linuxAccountsMgr{}).set()
+	}
+
+	if err := updateSSHConfig(enable, twofactor); err != nil {
+		logger.Errorf("Error updating SSH config: %v.\n", err)
```
Comment (adjackura) [inline GCEWindowsAgent/oslogin.go]: Ah, I missed this before, log lines inherently end in a new line correct? Wouldn't this add an extra one?

File: GCEWindowsAgent/accounts.go [hunk 11]
```diff
@@ -1,4 +1,4 @@
-//  Copyright 2017 Google Inc. All Rights Reserved.
+//  Copyright 2019 Google Inc. All Rights Reserved.
```
Comment (adjackura) [inline GCEWindowsAgent/accounts.go]: This should probably be linux_accounts.go since you have linuxAccountsMgr, would nixAccountsMgr be better?
